### DISCUSSION TITLE
Natively Support Web Components

### DISCUSSION THREAD

**Discussion by jonathantneal** [withastro/roadmap#57] [Proposal]
I wish Astro supported Web Components out-of-the-box.

```astro
---
// src/pages/index.astro
import MyCounter from '../components/MyCounter.js'
---
<MyCounter />
```

I wish Astro would detect when a component is an instance of `HTMLElement`.

```js
// src/components/MyCounter.js
export default class MyCounter extends HTMLElement {}
```

When it detects a web component, I wish Astro would automatically render it with a `<custom-name>`, and automatically add the script it exported from to the asset graph. If the constructor is not detected by `customElements.getName`, I wish Astro would generate an untaken name from the script filename and register it client-side with `customElements.define`. I wish Astro would register Web Components by default, without me writing any client directives.

```js
// generated /dist/assets/MyCounterRegister.js
import('/dist/assets/MyCounter.js').then(exports => customElements.define('html-mycounter', exports.default))
```

```html
<!-- generated /dist/index.html -->
<head>
  <script type="module" src="/dist/assets/MyCounterRegister.js"></script>
</head>
<body>
  <html-mycounter></html-mycounter>
</body>
```

**Comment by matthewp** [comment]
> I wish Astro would generate an untaken name from the script filename 

One issue with this part of your idea is that there wouldn't be a way to use that web component within another web component. For example, if I had another web component that was like:

```js
export default class extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `<html-mycounter></html-mycounter>`
  }
}
```

It wouldn't work, because the automatic naming code might have assigned it a different name. So this feature becomes only useful for top-level web components that aren't used outside of .astro files.

**Reply by jonathantneal** [reply]
Aye. That is a problem, albeit a problem faced by all Web Component™ frameworks. It’s the reason many frameworks ship their names hard-coded, which this would support:

```js
// src/components/MyCounter.js
export default class MyCounter extends HTMLElement {}

customElements.define('my-counter', MyCounter)
```

```js
// src/components/WrapMyCounter.js
export default class extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `<html-mycounter></html-mycounter>`
  }
}
```

Having read of consensus around `customElements.getName`, perhaps the client-side future may be:

```js
import MyCounter from './MyCounter.js'

// src/components/WrapMyCounter.js
export default class extends HTMLElement {
  connectedCallback() {
    const tag = customElements.getName(MyCounter)

    this.innerHTML = `<${tag}></${tag}>`
  }
}
```

However, in Astro, considering this known issue, and considering the kinds of React primitives being popularized, and considering our experience with those components being used in Astro, I suspect our users will prefer web components that let them bring their own LightDOM, rather than rely on client-side `innerHTML`.

**Reply by matthewp** [reply]
Web components use Shadow DOM, and this same problem occurs there as well. I'm having a hard time imagining a solution to this problem, actually. To only be able to use a web component that doesn't depend on other web components feels pretty limiting.

**Reply by jonathantneal** [reply]
The expectations of lit might be confusing what I’m asking for.

My suggestion does not forbid `customElements.define` in scripts as a side-effect, which is _one way_ to work around the issue of guaranteeing a tag name in server-side and client-side JS. I suggested another way, too, if someone wanted to polyfill `customElements.getName` in the same spirit as `<template shadowroot="open">`. And if component libraries did not export element constructors, then this proposal would not apply to them. They could continue to work as they work today.

**Reply by matthewp** [reply]
I guess what I'm getting at is, this part of the proposal won't work with most existing web components today. It would work with new components that are created to specifically work without knowing the name of the components they are using. Would you agree with that?

**Reply by jonathantneal** [reply]
>  It would work with new components that are created to specifically work without knowing the name of the components they are using.

Ah, thank you for sharing this. We should chat. This suggestion **does** work with components whose name is defined. Now, there’s a chance I’m still misunderstanding you, tho. My hope is that if we talk thru it, I can better understand your concern and/or do a better job explaining my suggestion. :)

**Reply by trusktr** [reply]
I've been wanting to try custom elements (with SSR) in Astro for a long time. Didn't know this thread existed until now. So glad y'all are pushing for it! 

I think what @matthewp might be getting at (or at least what I'm thinking), is that any elements written in this Astro-specific way that you're proposing @jonathantneal will not work anywhere else, only in Astro, locked into the system. If we copy them out to some other project, or one day want to move to Qwik.js or some other SSR framework, the elements won't work without needing some modification when written in some of the ways proposed.

I believe the best bet is for people who write Custom Elements (I much prefer that term over Web Components because React/Svelte/Vue/Solid components are all web components) to write them using the standards, and nothing more.

I believe this RFC should be scoped to the standard form of writing, just to get something out, and *then* if people really feel strongly about writing Custom Elements in a non-standard Astro-specific way, we can make a separate RFC to add that, or at least make the Custom Element renderer extensible so that the feature is easy to add (and someone can ship it as a library at least).

---

Regarding this:

> if I had another web component that was like:
> 
> ```js
> export default class extends HTMLElement {
>   connectedCallback() {
>     this.innerHTML = `<html-mycounter></html-mycounter>`
>   }
> }
> ```
> 
> It wouldn't work, because the automatic naming code might have assigned it a different name. So this feature becomes only useful for top-level web components that aren't used outside of .astro files.

Another proposed standard that would solve this issue is [Scoped Custom Element Registries](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md).

We'd write elements like this:

```js
import FarOut from 'somewhere'

const radicalRegistry = new CustomElementRegistry() // (Note, this would be shimmed for Astro SSR)

radicalRegistry.define('cowa-bunga', FarOut)

// Somebody else can name this element wherever they use it
export default class extends HTMLElement {
  constructor() {
    super()
    this.attachShadow({ mode: 'open', registry: radicalRegistry })
  }

  connectedCallback() {
    this.shadowRoot.innerHTML = `<cowa-bunga></cowa-bunga>`
  }
}
```

I personally like this more than the `customElements.getName()` idea, as we won't need the `` `<${tagName}>` `` syntax with scoped registries.

But these proposed standards are not reality yet, and I believe it would be good for the initial RFC to align with the minimal standard at first.

**Comment by RafidMuhymin** [comment]
What's about the following syntax using empty imports :
```js
---
import "../path/to/MyCounter.js";
---

<html>
  <body>
    <my-counter></my-counter>
  </body>
</html>
```

**Reply by jonathantneal** [reply]
That would only define `<my-counter>` if it were defined as a side-effect.

**Reply by jasikpark** [reply]
`import { my-counter } from "";` would be interesting if that were valid - using named exports would be nice

**Reply by trusktr** [reply]
I believe we should stick as much as possible to the existing ES and DOM standards, at least for the initial feature. (See https://github.com/withastro/roadmap/discussions/57#discussioncomment-6457515)

**Comment by RafidMuhymin** [comment]
And if Astro supports web components, then shouldn't we be able to import it in any component, for example: React, Preact, Svelte, Vue?

**Reply by trusktr** [reply]
Every framework (React, Preact, Svelte, Vue, Solid, etc) already supports using DOM elements inside of them, and Custom Elements are DOM elements, hence those frameworks can manipulate them (even outside of Astro).

So yeah, we can already import and use Custom Elements in any of those.

**However!** Astro might like to make the renderers for those components be a superset of a Custom Element renderer so that SSR for Custom Elements used inside any of those frameworks always works.

I.e. Custom Elements are already the common factor among all frameworks, so SSR for all the frameworks should implicitly also SSR any Custom Elements are used within them.

In the Astro [Framework Components](https://docs.astro.build/en/core-concepts/framework-components/#mixing-frameworks) doc, it says:

> Only **Astro** components (`.astro`) can contain components from multiple frameworks.

The exception to this rule should be noted, that Custom Elements are the only components that are already usable by default in any framework (and hence why we should ensure an SSR solution that always includes Custom Elements).

**Comment by matthewp** [comment]
> I wish Astro supported Web Components out-of-the-box.

I find this framing quite confusing. Astro does support web components out of the box:

```astro
<html>
<head><script src="/components/my-element.js"></script></head>
<body>
  <my-element></my-element>
</body>
</html>
```

I think what you mean is *automatically load web components in the client*, but it's unclear to me. I would recommend rephrasing this proposal around what you want to achieve. I strongly disagree that Astro doesn't support web components today.

----

About automatically adding JS, I think what you are alluding to in this proposal is one specific use-case for web components. It must be acknowledged that web components are not just used for client-side enhancement of HTML.

Many people use web components to build pages in the same manor as using other framework components like React. This is why the Lit renderer exists, because those users want to server-render their shadow DOM so that it can be seen before JavaScript runs in the client.

It wouldn't make sense to automatically load JavaScript for this use-case; the JavaScript isn't needed all of the time and is prone to the same problems of too-much-JS as any framework component can have.

I would like to see a solution proposed that allows automatically adding client-side JS that prevents accidentally over-adding, which is the reason we have the client directives.

**Reply by jonathantneal** [reply]
What I want is a little different than your example, and it’s also a little different than automatic script injection. I still want to update the title, and I need a little help. How would you phrase my wanting _this_?

```
---
// src/pages/index.astro
import MyCounter from '../components/MyCounter.js'
---
<MyCounter />
```

Is it "Natively Support Web Components as Components"?

**Reply by matthewp** [reply]
That seems like a renderer for a specific type of web component usage. Taking out the desire for automatic script loading, that should be achievable as a 3rd party renderer today.

**Reply by luwes** [reply]
We might need this functionality soon in our [Media Chrome docs](https://github.com/muxinc/media-chrome/pull/292)

I'm happy to help out w/ this. Quickly searching around I came on @thepassle's https://github.com/thepassle/custom-elements-ssr Astro plugin. Is this different than a 3rd party renderer?

@jonathantneal is the goal to get server side rendered custom elements w/ client hydration if needed?

**Reply by trusktr** [reply]
I'm a bit late to the conversation, but I think "Native support for Custom Elements" really means generically supporting all Custom Elements as with the current Lit renderer.

But the Lit renderer is only the beginning. We have more work to do, for example SSR of custom elements used anywhere in any other type of component (React, Svelte, Vue, Solid, etc) should just work. More description of that here:

https://github.com/withastro/roadmap/discussions/57#discussioncomment-6457560

**Reply by trusktr** [reply]
Maybe @thepassle has some ideas here
(he wrote [Server Side Rendering Vanilla Custom elements in Astro](https://dev.to/thepassle/server-side-rendering-vanilla-custom-elements-in-astro-5hgg) after having already taken Astro's Lit integration for a spin).

**Reply by luwes** [reply]
I have something working for our (Media Chrome) use case https://github.com/luwes/wesc but it should work for many custom elements. It works different than the Lit custom element renderer which evaluates every custom element tag separately I believe. We have a use case where we need to render html fragments with many custom elements together.

There's 2 meta frameworks that needed special code to SSR custom elements:
- React / Next.js
- [Astro](https://github.com/luwes/wesc/blob/main/src/astro/server.js)
  (thanks to @thepassle for his example plugin)

Would be great to get some feedback from Astro folks if this is best way to integrate with Astro.

**Reply by thepassle** [reply]
It's not entirely clear to me what exactly is wanted here. As I understand the thread there are two different topics at discussion:
- Jon's question about natively using web components _as_ Astro components
- An official Astro vanilla custom element SSR renderer - similarly to how Astro has an official renderer for Lit

I'm assuming we're mostly talking about the latter in this thread

Admittedly it's been a while since I wrote the vanilla renderer and I also haven't done much hacking on or with Astro recently, so I'm not sure how relevant my input here really is, but I'll say if anybody wants to pick up or take over maintenance of `custom-elements-ssr` I'd be more than happy to transfer it over. I think it'd make sense for Astro to ship a vanilla custom elements SSR renderer by default (considering there's also a Lit one), but thats up to the maintainers to decide, I can only state my preference :)

It worked fairly well for my usecase, however it should be mentioned that loading all of Linkedom for SSR purposes may not be the best if you're interested in high-perf usecases, and also Linkedom's maintainer has some weird personal feud wrt slots/shadowdom support in Linkedom, so methods like `slot.assignedNodes()` won't work. There's an open PR to add support for it, but apparently it's very timeconsuming to merge it. https://github.com/WebReflection/linkedom/pull/209

**Reply by matthewp** [reply]
I don't think we'll be taking up any new integrations unless there is huge number of users for them. The maintenance of the ones we already have is already too much. Finding new homes for existing ones is far more likely.

**Reply by thepassle** [reply]
I think that sounds fair
